### ISSUE TITLE
`CompletionException` in BSP server for `buildTarget/jvmRunEnvironment` request 

### ISSUE THREAD

**Issue by lefou** [com-lihaoyi/mill#2643]
@tgodzik wrote:
> I checked with 0.11.1 and the same exact thing seems to be happening:

```
[Trace - 04:37:26 PM] Received response 'buildTarget/jvmRunEnvironment - (16)' in 364ms
Result: null
Error: {
  "code": -32603,
  "message": "Internal error.",
  "data": "java.util.concurrent.CompletionException: java.lang.Exception: Failure during task evaluation: MillBsp.compile Compilation failed\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)\n\tat java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)\n\tat mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:670)\n\tat mill.bsp.worker.MillBuildServer.$anonfun$completable$1$adapted(MillBuildServer.scala:658)\n\tat scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)\n\tat java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)\nCaused by: java.lang.Exception: Failure during task evaluation: MillBsp.compile Compilation failed\n\tat mill.eval.Evaluator.$anonfun$evalOrThrow$default$1$1(Evaluator.scala:41)\n\tat mill.eval.EvaluatorImpl$EvalOrThrow.apply(EvaluatorImpl.scala:66)\n\tat mill.bsp.worker.MillBuildServer.$anonfun$completableTasks$1(MillBuildServer.scala:627)\n\tat mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:661)\n\t... 8 more\n"
}
```

_Originally posted by @tgodzik in https://github.com/com-lihaoyi/mill/issues/2499#issuecomment-1613301534_

**Comment by lefou** [comment]
Reformated stack trace

```
java.util.concurrent.CompletionException: java.lang.Exception: Failure during task evaluation: MillBsp.compile Compilation failed
    at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
    at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
    at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
    at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
    at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
    at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:670)
    at mill.bsp.worker.MillBuildServer.$anonfun$completable$1$adapted(MillBuildServer.scala:658)
    at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)
    at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:
    at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
    at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
    at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
    at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
    at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.Exception: Failure during task evaluation: MillBsp.compile Compilation failed
    at mill.eval.Evaluator.$anonfun$evalOrThrow$default$1$1(Evaluator.scala:41)
    at mill.eval.EvaluatorImpl$EvalOrThrow.apply(EvaluatorImpl.scala:66)
    at mill.bsp.worker.MillBuildServer.$anonfun$completableTasks$1(MillBuildServer.scala:627)
    at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:661)\n\t... 8 more
```

**Comment by carlosedp** [comment]
I'm debugging some stuck errors in Metals while using mill-bsp and added some logs and a capture to https://github.com/scalameta/metals/issues/4652.

I also see this CompletionException error in it.

**Comment by lefou** [comment]
There is another stack trace reported by Damian, https://matrix.to/#/!zXvttAKLPnvdnmiYsY:matrix.am/$124RiHjSmCeBGxv7U4GCApafH7L5-Q9FGA91kV92ib0?via=matrix.am&via=gitter.im&via=matrix.org

```
2023.07.03 09:35:38 INFO  BSP server: Jul 03, 2023 9:35:38 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
2023.07.03 09:35:38 INFO  BSP server: SEVERE: Internal error: java.nio.file.NoSuchFileException: /Users/damian/code/repos/github/DamianReeves/morphir-scala/out/generateScriptSources.super/mill/runner/MillBuildRootModule/generateScriptSources.dest
2023.07.03 09:35:38 INFO  BSP server: java.util.concurrent.CompletionException: java.nio.file.NoSuchFileException: /Users/damian/code/repos/github/DamianReeves/morphir-scala/out/generateScriptSources.super/mill/runner/MillBuildRootModule/generateScriptSources.dest
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:670)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1$adapted(MillBuildServer.scala:658)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1311)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1840)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1806)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
2023.07.03 09:35:38 INFO  BSP server: Caused by: java.nio.file.NoSuchFileException: /Users/damian/code/repos/github/DamianReeves/morphir-scala/out/generateScriptSources.super/mill/runner/MillBuildRootModule/generateScriptSources.dest
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:248)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:105)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.nio.file.Files.delete(Files.java:1152)
2023.07.03 09:35:38 INFO  BSP server: 	at os.remove$all$.apply(FileOps.scala:331)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.GroupEvaluator.evaluateGroupCached(GroupEvaluator.scala:120)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.GroupEvaluator.evaluateGroupCached$(GroupEvaluator.scala:35)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorImpl.evaluateGroupCached(EvaluatorImpl.scala:17)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.$anonfun$evaluate0$2(EvaluatorCore.scala:116)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.ExecutionContexts$RunNow$.execute(ExecutionContexts.scala:13)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:312)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$DefaultPromise.map(Promise.scala:182)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.$anonfun$evaluate0$1(EvaluatorCore.scala:92)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.$anonfun$evaluate0$1$adapted(EvaluatorCore.scala:90)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.collection.immutable.VectorStatics$.foreachRec(Vector.scala:2122)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.collection.immutable.VectorStatics$.foreachRec(Vector.scala:2128)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.collection.immutable.Vector.foreach(Vector.scala:2128)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.evaluate0(EvaluatorCore.scala:90)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.$anonfun$evaluate$1(EvaluatorCore.scala:43)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.evaluate(EvaluatorCore.scala:34)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorCore.evaluate$(EvaluatorCore.scala:26)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorImpl.evaluate(EvaluatorImpl.scala:17)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.eval.EvaluatorImpl$EvalOrThrow.apply(EvaluatorImpl.scala:64)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completableTasks$1(MillBuildServer.scala:627)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:661)
2023.07.03 09:35:38 INFO  BSP server: 	... 8 more
2023.07.03 09:35:38 INFO  BSP server: 
2023.07.03 09:35:38 INFO  mill-bsp should support `buildTarget/jvmRunEnvironment`, but it fails.
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Internal error.
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1589)

2023.07.03 09:35:38 INFO  BSP server: Jul 03, 2023 9:35:39 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
2023.07.03 09:35:38 INFO  BSP server: SEVERE: Internal error: java.lang.ClassNotFoundException: munit.Framework
2023.07.03 09:35:38 INFO  BSP server: java.util.concurrent.CompletionException: java.lang.ClassNotFoundException: munit.Framework
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:670)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1$adapted(MillBuildServer.scala:658)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1311)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1840)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1806)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
2023.07.03 09:35:39 INFO  Failed to run request with params ScalaTestClassesResult
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Internal error.
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.lang.Thread.run(Thread.java:1589)

2023.07.03 09:35:38 INFO  BSP server: Caused by: java.lang.ClassNotFoundException: munit.Framework
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.api.ClassLoader$$anon$1.findClass(ClassLoader.scala:35)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:588)
2023.07.03 09:35:38 INFO  BSP server: 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.testrunner.Framework$.framework(Framework.scala:9)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillScalaBuildServer.$anonfun$buildTargetScalaTestClasses$9(MillScalaBuildServer.scala:113)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.util.Jvm$.inprocess(Jvm.scala:254)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillScalaBuildServer.$anonfun$buildTargetScalaTestClasses$7(MillScalaBuildServer.scala:121)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completableTasks$3(MillBuildServer.scala:628)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.collection.immutable.Vector2.map(Vector.scala:2152)
2023.07.03 09:35:38 INFO  BSP server: 	at scala.collection.immutable.Vector2.map(Vector.scala:441)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completableTasks$1(MillBuildServer.scala:628)
2023.07.03 09:35:38 INFO  BSP server: 	at mill.bsp.worker.MillBuildServer.$anonfun$completable$1(MillBuildServer.scala:661)
2023.07.03 09:35:38 INFO  BSP server: 	... 8 more
2023.07.03 09:35:38 INFO  BSP server: 
```

As I spotted this line:

```
Caused by: java.nio.file.NoSuchFileException: /Users/damian/code/repos/github/DamianReeves/morphir-scala/out/generateScriptSources.super/mill/runner/MillBuildRootModule/generateScriptSources.dest
```

I think it might be related to this comment: https://github.com/com-lihaoyi/mill/issues/2629#issuecomment-1612975150

> I have some suspicion, since I currently try to adapt `GenIdea` to the new bootstrap process. It's unvalidated, but maybe we should check it. It could be, we sometimes use the wrong evaluator when evaluating tasks. E.g. in BSP, we search for all root modules and transitively collect other modules, then we only keep them for later resolution of sources, resources and classpaths. But depending on the root module, we need to use different evaluators (evaluator configs, e.g. root directory, classpath, etc.). I'm not sure this is always the case.

**Comment by DamianReeves** [comment]
Has anyone else made any progress on this issue. This has become a huge problem for my team and I'm at the point of thinking of switching our rather large Mill build back to sbt so teammates can get working IDEs/ VSCode..

**Comment by lefou** [comment]
> Has anyone else made any progress on this issue. This has become a huge problem for my team and I'm at the point of thinking of switching our rather large Mill build back to sbt so teammates can get working IDEs/ VSCode..

I understand your frustration. I already posted my idea of the issue. The envisioned fix isn't trivial though. Of course, anybody is free to work on whatever they want and extra pressure isn't much rewarding.

Before switching to a completely different build tool, which probably comes with it's own issues, you should consider just downgrading to Mill 0.10.12. BSP support might not be as good as it could be in Mill 0.11, but if the latter is broken for you, it might do. You could also give IntelliJ IDEA a try, which works very well for me and also others, also in the Community Edition. Donating some developer time towards a fix might be another option.

**Comment by lefou** [comment]
> I don't think there's a way to check what project a module depends on. In theory, each set of modules should be in a totally different classloader and one depth in MillBuildBootstrap resulting in an evaluator with one root module folder
> 
> Yeah it definitely sounds like some of the meta build modules are being evaluated using the wrong evaluator. And this seems to happen only when using mill-bsp. IIRC we cache the evaluator in some parts of the mill-bsp codebase; could it be that cache is being kept around too long when it should instead be invalidated?

I think it's rather the fact that we collect all modules, including the ones of the meta-build, when building up the BSP state (we do this to represent all modules, also the meta-build in the IDE). While doing so, we drop the extra information of the project root folder, which would be needed to calculate the correct destination folder (`EvaluatorPaths.resolveDestPaths` takes a `workspacePath`).

**Comment by tgodzik** [comment]
> You could also give IntelliJ IDEA a try, which works very well for me and also others, also in the free Community Edition.

I think Intellij doesn't use some specific endpoints that we do, which is why it's working better in this particular case.

**Comment by lefou** [comment]
> > You could also give IntelliJ IDEA a try, which works very well for me and also others, also in the free Community Edition.
> 
> I think Intellij doesn't use some specific endpoints that we do, which is why it's working better in this particular case.

Oh, I just realize that you can IDEA as BSP client, or by generating the config with `mill.scalalib.GenIdea/idea`. I for myself use the latter.

**Comment by lefou** [comment]
@lihaoyi It would be nice if we could encode the meta-build-ness in something like the `foreignSegments` or if we could reuse the `foreignSegments`, as it is probably impossible to have a foreign module and a meta-build module in the same location for the same project.

**Comment by lihaoyi** [comment]
It's definitely possible. But if the problem is mill-bsp mixing the modules together and losing the distinction between stages, maybe we should just keep the information in mill-bsp instead in some dictionary or something? Since the vast majority of Mill code doesn't need to know about this as far as I can tell, putting it on the Module object itself may be exposing it unnecessarily broadly

**Comment by lefou** [comment]
> It's definitely possible. But if the problem is mill-bsp mixing the modules together and losing the distinction between stages, maybe we should just keep the information in mill-bsp instead in some dictionary or something? Since the vast majority of Mill code doesn't need to know about this as far as I can tell, putting it on the Module object itself may be exposing it unnecessarily broadly

Now, we actually need those info to have a proper unique module name. We also need it in `GenIdea`. We could share it between GenIdea and BSP, but time has shown that information directly in the core has better maintenance than IDE integration. And good IDE integration is nowadays a must. We probably also need this info later, when we try to make meta-build runnable/inspectable from the CLI.

**Comment by DamianReeves** [comment]
> @DamianReeves I understand your frustration. I already posted my idea of the issue. The envisioned fix isn't trivial though. Of course, anybody is free to work on whatever they want and extra pressure isn't much rewarding.

Sorry I wasn't meaning to add pressure to you, and more communicating the pressure I am experiencing.
I would really love to try and debug/offer a fix to this. Maybe with some guidance I can do so. We happen to be seeing other BSP errors which I'd love to debug and try and target where we are experiencing it. I just don't have experience debugging/working with BSP.

Sorry again. I was intending to raise urgency (i.e. this is important not a nice to have), not pressure.

**Comment by DamianReeves** [comment]
> Oh, I just realized that it was ambiguous, as you can use IDEA as BSP client, or by generating the config with `mill.scalalib.GenIdea/idea`. I for myself use the latter.

We've seen some issues here as well, or my teammate has. I am able to work, but others seem to have issues.

**Comment by DamianReeves** [comment]
Given this discussion, is a workaround for me to stop using meta-build?

**Comment by lefou** [comment]
> Given this discussion, is a workaround for me to stop using meta-build?

This is definitely worth a try.

The situation with BSP is currently, that we more or less fly blindfolded, as we don't have proper integration testing for it in Mill. Other BSP servers have it, Metals also runs some, and there is also a bsp-testkit(2), but I still haven't found the time to adapt it to Mill. Once we have integration testing for BSP, we could make and keep it more robust, faster and easier. Currently, it's manual testing and luck to find issues, which is not very rewarding, quite the opposite. Adding some integration testing for our BSP server to assert expected BSP responses for given projects would be a great contribution. I think there is almost no knowledge about Mill internals needed, as it's mostly protocol testing.

**Comment by DamianReeves** [comment]
Where can I find this `bsp-testkit`?

**Comment by lefou** [comment]
* bsp-testkit2: https://github.com/build-server-protocol/bsp-testkit2
* bsp-testkit: https://github.com/build-server-protocol/build-server-protocol/tree/master/testkit
* Metals MillServerSuite: https://github.com/scalameta/metals/tree/main/tests/slow/src/test/scala/tests/mill

**Comment by lefou** [comment]
We have merged #2692, which should fix this issue. It would be nice if you could confirm the fix. I started CI job https://github.com/com-lihaoyi/mill/actions/runs/5864992453 to publish a snapshot release

**Comment by tgodzik** [comment]
> We have merged #2692, which should fix this issue. It would be nice if you could confirm the fix. I started CI job https://github.com/com-lihaoyi/mill/actions/runs/5864992453 to publish a snapshot release.
> 
> /cc @tgodzik @DamianReeves

Looks to be working! I need to do some more work to make run work in metals, but that should be ready soon. The only issue is that it will still throw an exception when the code doesn't compile, which will show up in the logs and might be annoying for users. Ideally, it would just return last jvmEnvironment especially that this shouldn't be influenced by regular compilation.

**Comment by tgodzik** [comment]
Ok, the PR is ready https://github.com/scalameta/metals/pull/5566 though I need to refactor things at some point since we get a lot of different options when things can be run (via Dap or via build tool itself or via client)
